Ocp migration new

apinnick opened the pull request:
Preview: http://file.tlv.redhat.com/~apinnick/ocp-migration-new/migration/migrating-openshift-3-to-4.html

gildub commented:
@apinnick, 

First off, kudos for this hard work!

In the "Using CPMA" - "Procedure" section we actually need to add between 1. and 2. (either as 1 subsection or pushing 2. down) a paragraph about ssh permission and also grant privileges.

Please see https://github.com/fusor/cpma/#authentication-and-authorization for more details and/or don't hesitate to ask me.

bergerhoffer commented on modules/migration-creating-migration-plan-cam.adoc:
```diff
@@ -0,0 +1,24 @@
+// migration/migrating_openshift_3_to_4/migrating-openshift-3-to-4.adoc
+[id='migration-creating-migration-plan-cam_{context}']
+= Creating a migration plan in the CAM web console
+
```
Same comment about adding an intro statement if possible.

bergerhoffer commented on modules/migration-creating-migration-plan-cam.adoc:
```diff
@@ -0,0 +1,24 @@
+
+.Prerequisites
+
+* Source and target clusters added to the CAM web console
```
Need to either say "have been added" or reword to something like "You have added the source and target clusters to the CAM web console."

apinnick replied:
Done. Did the same for the storage.

bergerhoffer commented on modules/migration-custom-resources.adoc:
```diff
@@ -0,0 +1,40 @@
+
+image::migration-architecture.png[migration architecture diagram]
+
+<1> link:https://github.com/fusor/mig-controller/blob/master/pkg/apis/migration/v1alpha1/migcluster_types.go[MigCluster] (configuration, CAM cluster): Cluster definition
```
@sferich888 Are you okay with these references to these GitHub repos?

bergerhoffer commented on modules/migration-restic-timeout.adoc:
```diff
@@ -0,0 +1,28 @@
+
+.Procedure
+
+. In the web console, navigate to *Catalog* -> *Installed Operators*.
```
Can you be clear if this is the CAM web console or the {product-title} web console?

bergerhoffer commented on modules/migration-restic-timeout.adoc:
```diff
@@ -0,0 +1,28 @@
+. In the *MigrationController* tab, click *migration-controller*.
+. In the *YAML* tab, update the following parameter value:
++
+----
```
Add [source,yaml] above this.

bergerhoffer commented on modules/migration-installing-cpma.adoc:
```diff
@@ -0,0 +1,14 @@
+. Select *CPMA 1.0 for RHEL 7* from the *Version* list. The same binary will work on RHEL 7 or RHEL 8.
+. Select *x86_64*, which is the default, from the *Architecture* list.
++
+The choices for downloading CPMA for each platform (Linux, Windows, and Mac) populate.
```
Mac -> macOS

bergerhoffer commented on modules/migration-installing-migration-operator-manually.adoc:
```diff
@@ -0,0 +1,63 @@
+
+.Procedure
+
+. Download the *_operator.yml_* file:
```
File names should be in backticks. (See other instances in this file and throughout PR)

apinnick replied:
Duplicate comment. I've left the other one open.

bergerhoffer commented on modules/migration-installing-migration-operator-manually.adoc:
```diff
@@ -0,0 +1,63 @@
+$ podman cp $(podman create registry-redhat.io/rhcam/openshift-migration-rhel7-operator:v1.0 ):/controller-4.yml ./
+----
++
+The *_controller-4.yml_* file installs the CAM on the {product-title} 4 cluster by default. If you do not wish to install the CAM on this cluster, update the values of the following parameters:
```
wish -> want

bergerhoffer commented on modules/migration-installing-migration-operator-manually.adoc:
```diff
@@ -0,0 +1,63 @@
++
+The *_controller-4.yml_* file installs the CAM on the {product-title} 4 cluster by default. If you do not wish to install the CAM on this cluster, update the values of the following parameters:
++
+----
```
Add [source,yaml]

gildub commented on modules/migration-using-cpma.adoc:
```diff
@@ -3,21 +3,26 @@
+* The {product-title} 3 cluster must be updated to the latest synchronous release.
+* The environment health check must be run on the {product-title} 3 cluster to confirm that there are no diagnostic errors or warnings.
+* The {product-title} version must be 3.7, 3.9, 3.10, or 3.11.
+* If you are running CPMA in remote mode, the SSH user must have sudo permissions to read the configuration files. If this is not possible, you must copy the files locally and run CPMA in local mode.
```
I think that SSH user details and Openshift user privileges below both belong to the .Procedure section as setup.

apinnick replied:
What do you see as the difference between prerequisites and setup? 
SSH user details and ocp user privileges are more of a state than an action.

gildub replied:
One is the `Prerequisites` is about what's needed prior to do anything before downloading (or using) CPMA.

The second, `Setup` are the details required as part of getting CPMA working.

gildub commented on modules/migration-using-cpma.adoc:
```diff
@@ -3,21 +3,26 @@
+* The environment health check must be run on the {product-title} 3 cluster to confirm that there are no diagnostic errors or warnings.
+* The {product-title} version must be 3.7, 3.9, 3.10, or 3.11.
+* If you are running CPMA in remote mode, the SSH user must have sudo permissions to read the configuration files. If this is not possible, you must copy the files locally and run CPMA in local mode.
+* You must have `cluster-admin` privileges.
```
What about proving the grant command too?
`oc adm policy add-cluster-role-to-user cluster-admin <username>`

apinnick replied:
I didn't add this line because I didn't see it as relevant to this procedure. If someone needs to grant cluster-admin privileges to the user, it is well documented elsewhere.

gildub replied:
Fair enough but what about adding a link to the documentation then?

apinnick replied:
OK, will do.

bergerhoffer commented on modules/migration-installing-migration-operator-olm.adoc:
```diff
@@ -0,0 +1,56 @@
+
+.Procedure
+
+. Click *Administration* -> *Namespaces*.
```
Again be specific about which web console this is?

bergerhoffer commented on modules/migration-installing-migration-operator-olm.adoc:
```diff
@@ -0,0 +1,56 @@
+
+. On the *Installed Operators* page:
+
+.. Under *Provided APIs*, click *View 12 more...*.
```
I wouldn't call out "12 more". This might change.

apinnick replied:
Unfortunately, we have to do this right now. There's no other way to get to the *Create New* button. Blame the OperatorHub designers....

bergerhoffer commented on modules/migration-installing-migration-operator-olm.adoc:
```diff
@@ -0,0 +1,56 @@
+
+.. If you are installing the CAM Operator on an {product-title} _4.1_ cluster, add the following parameter:
++
+----
```
Add [source,yaml]

bergerhoffer commented on modules/migration-installing-migration-operator-olm.adoc:
```diff
@@ -0,0 +1,56 @@
+
+.. If you are installing the CAM Operator on an {product-title} 4 _source_ cluster, update the following parameter values so that the migration controller and CAM web console are not installed:
++
+----
```
[source,yaml]

bergerhoffer commented on modules/migration-installing-migration-operator-olm.adoc:
```diff
@@ -0,0 +1,56 @@
+.. Click *Workloads* -> *Pods* to verify that the controller manager, migration UI, Restic, and Velero pods are running.
+.. Click *Networking* -> *Routes*. In the `openshift-migration` namespace, the CAM host URL is the URL listed under *Location* if the CAM is running on this cluster. You will use the CAM host URL to configure cross-origin resource sharing on the other clusters and to launch the CAM web console.
+
+. If you did not install the migration controller, click *Workloads* -> *Pods* to verify that the Rustic and Velero pods are running.
```
Throughout the PR, "pods" -> "Pods"

apinnick replied:
"Pods" is capitalized? Why?! It's a generic term. 

Did a quick search in openshift/modules. 
"pods" - 923 times in 272 files.
"Pods" - 354 times in 153 files.

But I'll change it, since maybe this is a recent style decision. Seems odd to me.

apinnick replied:
Is this only for named pods (e.g., Restic Pods) or for all pods (e.g., static Pods)?

bergerhoffer replied:
All instances of Pod/Pods should be uppercase according to our guidelines: https://github.com/openshift/openshift-docs/blob/master/contributing_to_docs/term_glossary.adoc#pod

I agree, I don't like this guidance either, but it was decided at least a year ago (before I joined the OpenShift docs team) so I follow it.

bergerhoffer commented on modules/migration-known-issues.adoc:
```diff
@@ -0,0 +1,27 @@
+[id='migration-known-issues_{context}']
+= Known issues
+
+The following issues will be addressed in a future release:
```
I'd be careful about saying this, these issues might not ever get fixed.

apinnick replied:
What do you suggest I use as a lead-in? "The following are known issues:"?

bergerhoffer replied:
You could do anything like:

* This release has the following known issues:
* Review the following known issues:

bergerhoffer commented on modules/migration-launching-cam.adoc:
```diff
@@ -0,0 +1,22 @@
+
+.Procedure
+
+. If you are using self-signed certificates, launch a browser and accept the CA certificates manually for the following:
```
Do you have to try to navigate to the CAM web console before this step?

apinnick replied:
Nope

bergerhoffer commented on modules/migration-launching-cam.adoc:
```diff
@@ -0,0 +1,22 @@
+* Source cluster's OAuth server, for example, `https://_<master1.cluster.com>_/.well-known/oauth-authorization-server`
+* Source cluster's API server, for example, `https://_<master1.cluster.com>_/api/v1/namespaces`
+
+. Navigate to the CAM web console.
```
Can you give users any tips on what the URL for the CAM web console is, or where to find it?

apinnick replied:
It's mentioned twice before, in both the CORS config sections, but I guess it doesn't hurt to hit them over the head with it. Will add.

bergerhoffer commented on modules/migration-obtaining-service-account-token-camt.adoc:
```diff
@@ -0,0 +1,22 @@
+// Module included in the following assemblies:
+//
+// migration/migrating_openshift_3_to_4/migrating-openshift-3-to-4.adoc
```
I don't see this file included in the assembly anywhere, should it be? Or was this step included into the "Adding a cluster to the CAM web console" procedure and this file should be removed?

apinnick replied:
Good catch. I've removed the file.

bergerhoffer commented on modules/migration-running-migration-plan-cam.adoc:
```diff
@@ -0,0 +1,27 @@
+. Select a migration plan and click *Stage* (copies the data from the source cluster to the target cluster without stopping the application) or *Migrate* (stops the application on the source cluster and moves the data to the target cluster).
+
+. Optionally, if you click *Migrate*, you can select *Do not stop applications on the source cluster during migration*.
+. Click *Migrate*.
```
This procedure is a bit confusing. Maybe it would make more sense with the CAM web console in front of me, but are the steps really to either 1). select a plan, click Stage, then click Migrate or to 2). select a plan, click Migrate, then click Migrate again? 

Seems confusing with multiple "Migrate" options. Maybe separate Stage/Migrate from the first step to make them clearer?

apinnick replied:
Not sure how to explain this. "Stage" and "Migrate" are separate steps but they do not require separate procedures, IMO. The differences between the procedures are explained in the "About CAM" section. The user might choose to stage a migration plan more than once, but would only migrate with the migration plan once.

I'll see whether I can make it clearer, without duplicating the procedure.

bergerhoffer commented on modules/migration-running-migration-plan-cam.adoc:
```diff
@@ -0,0 +1,27 @@
+. Optionally, if you click *Migrate*, you can select *Do not stop applications on the source cluster during migration*.
+. Click *Migrate*.
+
+. When the migration is complete, verify that the application migrated successfully in the target cluster's web console:
```
I would be explicit that this is now the {product-title} cluster.

bergerhoffer commented on modules/migration-understanding-cpma.adoc:
```diff
@@ -0,0 +1,41 @@
+[id='migration-understanding-cpma_{context}']
+= Understanding the Control Plane Migration Assistant
+
+The Control Plane Migration Assistant (CPMA) is a CLI-based tool that assists you in migrating the control plane from {product-title} 3.x to 4.x. The CPMA processes {product-title} 3.x configuration files and generates Custom Resource (CR) manifest files, which are consumed by {product-title} 4.x operators.
```
operators -> Operators

bergerhoffer commented on modules/migration-understanding-cpma.adoc:
```diff
@@ -0,0 +1,41 @@
+
+CPMA uses the Kubernetes and {product-title} APIs to access the following configuration files on an {product-title} 3 cluster:
+
+* Master configuration file (default: *_/etc/origin/master/master-config.yaml_*)
```
Same in this file about using backticks for file names

bergerhoffer commented on modules/migration-understanding-cpma.adoc:
```diff
@@ -0,0 +1,41 @@
++
+[NOTE]
+====
+If you are using an unsigned certificate, you must link:https://docs.openshift.com/container-platform/4.1/authentication/certificates/api-server.html#add-named-api-server_api-server-certificates[add the API server certificate].
```
You can't link here. We post the OpenShift docs both on docs.openshift.com and the customer portal, so this link would be incorrect on the portal. So unfortunately we just can't link at all to docs from modules right now.

apinnick replied:
So how do you link to different parts of your documentation?

apinnick replied:
Are relative links OK?

link:../openshift_images/image-streams-manage.html#images-imagestreams-update-tag_image-streams-managing[update the `imagestreamtags` references]

apinnick replied:
Never mind -- I see that relative links are OK - https://github.com/openshift/openshift-docs/blob/dd2a3461de1e2ddc61b98c7bb3186e68950edd4e/contributing_to_docs/doc_guidelines.adoc#links-hyperlinks-and-cross-references

bergerhoffer replied:
You can't link to other parts of the docs when you're inside a module. You can only use xrefs as long as you're in an assembly.

Your link for the guidelines is outdated. See: https://github.com/openshift/openshift-docs/blob/master/contributing_to_docs/doc_guidelines.adoc#links-hyperlinks-and-cross-references

bergerhoffer commented on modules/migration-using-cpma.adoc:
```diff
@@ -0,0 +1,92 @@
+// migration/migrating-openshift-3-to-4.adoc
+[id='migration-using-cpma_{context}']
+= Using CPMA
+
```
Intro sentence

bergerhoffer commented on modules/migration-using-cpma.adoc:
```diff
@@ -0,0 +1,92 @@
+
+.Prerequisites
+
+The following prerequisites apply to the {product-title} 3 cluster:
```
I don't think it's in the template to include a sentence like this. I would just be specific in each of your prereqs that it it applies to the 3.x cluster.

bergerhoffer commented on modules/migration-using-cpma.adoc:
```diff
@@ -0,0 +1,92 @@
+. Run CPMA. Each prompt requires you to provide input, as in the following example, which produces a report without the CR manifests:
++
+----
+$ ./cpma --manifests=false
```
This seems to assume that you're in the directory you downloaded CPMA to. I'd add a step about navigating to that directory.

apinnick replied:
No, it doesn't. `./cpma` is an absolute path, so you can run it from anywhere.

apinnick replied:
OK, It's just an rpm, so I'll add a yum install command. It will create the /cpma directory.

bergerhoffer replied:
Is "cpma" a directory or a command? The way it's used above, it's an executable file.

And the "." in "./cpma" means "in the current directory", so it's a relative path.

bergerhoffer commented on modules/migration-using-cpma.adoc:
```diff
@@ -0,0 +1,92 @@
++
+CPMA creates the following files and directory in the current directory if you did not specify an output directory:
+
+* *_cpma.yaml_* file: Configuration options that you provided when you ran CPMA
```
backticks for all these file names

bergerhoffer commented on modules/migration-installing-cpma.adoc:
```diff
@@ -0,0 +1,14 @@
+// migration/migrating-openshift-3-to-4.adoc
+[id='migration-installing-cpma_{context}']
+= Installing CPMA
+
```
Intro

bergerhoffer commented:
@apinnick Can you also specify what version(s) this is for? Is it for 4.2+, or is it for 4.1 as well?

apinnick commented:
@bergerhoffer It's for 4.2+

apinnick commented on _topic_map.yml:
```diff
@@ -1235,6 +1235,13 @@ Topics:
 - Name: Release Notes
   File: serverless-release-notes
 ---
+Name: Migration
```
@vikram-redhat Please comment. You told me where to put the Migration bucket.

vikram-redhat replied:
I think @bergerhoffer's suggestion is valid. We should move it.

apinnick commented on modules/migration-understanding-cam.adoc:
```diff
@@ -0,0 +1,32 @@
+
+* *Copying* copies the data in a source cluster's PV to an S3 repository and then restores it on a newly created PV, with similar characteristics, in the target cluster. You can change the storage class during migration to any storage class that is available on the target cluster.
++
+image::migration-PV-copy.png[Copy data in a source cluster's PV]
```
They were done by Jess Schaefer.

bergerhoffer replied:
My apologies. I was expecting the color to be green like: https://docs.openshift.com/container-platform/4.2/architecture/architecture.html

apinnick commented on migration/migrating-openshift-3-to-4.adoc:
```diff
@@ -0,0 +1,92 @@
+
+toc::[]
+
+You can migrate application workloads from {product-title} 3.7 to a later version with the Cluster Application Migration (CAM) tool. The CAM enables you to control the migration and to minimize application downtime. The CAM provides a web console and an API, based on Kubernetes custom resources, for migrating stateful application workloads, at the granularity of a namespace, from a source to a ta ...
```
> This first sentence makes it sound like you can only migrate from version 3.7 exactly. Suggest rewording
> 
> `You can migrate application workloads from {product-title} 3.7 and later to {product-title} 4.x with the...`

Actually, 3,7+ can be located to any later version, not just 4.x. I'll add 'and later' to 3.7. But that is closely followed by "to a later version". Will have to think about that.

bergerhoffer replied:
So my thoughts are that, even if you can use the tool for 3.x -> 3.x, these docs are living in the OCP 4.2 doc set. So we should assume that's the version users are on (or going to).

I don't think we'd want to get into users using these docs when working with 3.x -> 3.x migrations.

apinnick replied:
True. It's a problem that has never really been resolved. The doc is about migrating to 4.2 but they want users to know that it's possible to migrate to 3.x. I'll see whether I can put that into a note somewhere, so that the 3 > 4.2 path is clearer.

apinnick commented on migration/migrating-openshift-3-to-4.adoc:
```diff
@@ -0,0 +1,92 @@
+The CAM Operator installed on all clusters:
+
+* {product-title} 3.x: You must install the CAM Operator manually, because OLM is not supported.
+* {product-title} 4.x: You can install the CAM Operator with OLM.
```
They can do it manually but we don't want them to. In the manual installation section, we say that they can install it manually on ocp 4

apinnick commented on migration/migrating-openshift-3-to-4.adoc:
```diff
@@ -0,0 +1,92 @@
+
+include::modules/migration-configuring-cors-3.adoc[leveloffset=+2]
+
+You can now verify xref:migration-verifying-cors_{context}[cross-origin resource sharing].
```
> @apinnick I'm not sure whether you were ready for a peer review, but I heard that this is going to go out this week, and it needed to be peer reviewed and approved after addressing feedback before it could get merged. So I wanted to make sure we had enough time.
> 

Not ready for peer review. Still need some input on Noobaa.

> Let me know if you have any questions on the feedback. Thanks!
> 
> One other item that I couldn't comment directly inline about is that there are darkcircle*.png files, that don't seem to be used. Can you remove these?

They're used in the CR section. I just added them.

apinnick replied:
I put it in so that they wouldn't skip the verification step by accident if they've just configured CORS on OCP 3.

bergerhoffer replied:
I might just work it in as the last step of the procedure. But again, it's fine to leave it as is.

apinnick replied:
Oh, duh! Why didn't I think of that? :-) Probably because I got the verification step so late that it was simpler to stick it in a separate module. I added the step to the configuration procedures (both modules, but we're not going to use the CORS on ocp 4 cluster until next release). Thanks!

apinnick commented:
> @apinnick I'm not sure whether you were ready for a peer review, but I heard that this is going to go out this week, and it needed to be peer reviewed and approved after addressing feedback before it could get merged. So I wanted to make sure we had enough time.
> 

Not ready for peer review. Still need some input on Noobaa.

> Let me know if you have any questions on the feedback. Thanks!
> 
> One other item that I couldn't comment directly inline about is that there are darkcircle*.png files, that don't seem to be used. Can you remove these?

They're used in the CR section. I just added them.

jwmatthews commented on migration/migrating-openshift-3-to-4.adoc:
```diff
@@ -0,0 +1,92 @@
+* You must have `podman` installed.
+* The source cluster(s) must be {product-title} 3.7 or later.
+* The target cluster must be {product-title} 3.9 or later.
+* The replication repository object storage must support the S3 API, for example, NooBaa, AWS S3, and generic S3 storage providers.
```
Consider adding a blurb like below to point out a link for an example of leveraging the upstream NooBaa for a local S3 solution.  (Once OCS 4 is GA we will want to update this and refer to a procedure for using a supported downstream build, for time being recommendation is to proceed with unsupported upstream example).

"Examples of deploying a local S3 provider leveraging the upstream NooBaa project or using AWS S3 may be found at: https://github.com/fusor/mig-operator/blob/release-1.0/docs/usage/ObjectStorage.md"

apinnick replied:
@jwmatthews Is it OK to have links to github? We also have them for the migration CRs. I thought they were problematic.

sferich888 replied:
In this case yes. Support can approve links to GitHub, so long as they are used for example and non critical items.

apinnick replied:
Looks like it's OK if they're examples, so I'll add something like this.

bergerhoffer replied:
Thanks @sferich888!

XinRedhat commented on modules/migration-understanding-cam.adoc:
```diff
@@ -0,0 +1,32 @@
++
+image::migration-PV-copy.png[Copy data in a source cluster's PV]
+
+* *Moving* unmounts a remote volume (for example, NFS) from the source cluster, creates a PV resource on the target cluster pointing to the remote volume, and then mounts the remote volume on the target cluster. Applications running on the target cluster use the same remote volume that the source cluster was using. (The remote volume must be accessible to the source and target clusters.)
```
For **Moving** option, I remember it just supports migrating data NFS -> NFS storage so far. Do we need to explicitly address it?

apinnick replied:
> For **Moving** option, I remember it just supports migrating data NFS -> NFS storage so far. Do we need to explicitly address it?

@jwmatthews Can you respond?

jwmatthews replied:
The move functionality is not limited to just NFS.
NFS is most common use case of this and has been the focus of testing.
In the future we see other usages.  (Development has verified of usage of move for external gluster PVs for example).

apinnick commented on modules/migration-configuring-cors-3.adoc:
```diff
@@ -0,0 +1,55 @@
+----
+
+. Log in to the {product-title} 3 cluster.
+. Add the CAM host URL to the `corsAllowedOrigins` stanza in the *_/etc/origin/master/master-config.yaml_* configuration file:
```
According to the the OpenShift doc guidelines, file names and directories are marked as *_filename_*.

https://github.com/openshift/openshift-docs/blob/dd2a3461de1e2ddc61b98c7bb3186e68950edd4e/contributing_to_docs/doc_guidelines.adoc#quick-markup-reference

bergerhoffer replied:
I'm not sure where that link came from, but it's out of date. The correct one is in the 'master' branch:

https://github.com/openshift/openshift-docs/blob/master/contributing_to_docs/doc_guidelines.adoc#quick-markup-reference

apinnick commented on modules/migration-adding-replication-repository-to-cam.adoc:
```diff
@@ -0,0 +1,28 @@
+
+.Procedure
+
+. In the *Replication repositories* section, click *Add replication repository*.
```
Good point. I always performed the steps sequentially but a user might not.

apinnick commented on modules/migration-launching-cam.adoc:
```diff
@@ -0,0 +1,22 @@
+
+. If you are using self-signed certificates, launch a browser and accept the CA certificates manually for the following:
+
+* CAM host's OAuth and API server, for example, `https://_<migration-openshift-migration.apps.cluster.com>_:6443/.well-known/oauth-authorization-server`
```
That markup seems to be used a great deal in ocp docs, for example, modules/identity-provider-about-request-header.adoc

The guidelines say to use variable markup. Does that mean `<project>` or <project> if the variable is part of a URL?

bergerhoffer replied:
Yeah usually we would use <cluster_name> or something anywhere they'd need to replace the value. But like I said you can always use an actual example like what you already have and then just mention to be sure to replace it with your actual value.

apinnick commented on modules/migration-configuring-cors-4.adoc:
```diff
@@ -0,0 +1,68 @@
+
+If you installed the migration controller on your {product-title} 4 cluster, the cluster's resources are modified by the CAM Operator and you do not need to configure cross-origin resource sharing. If you did not install the migration controller on this cluster (for example, if you are using another cluster to host the CAM), you must configure cross-origin resource sharing manually.
+
+To access the API server of an {product-title} 4 cluster from a web application (in this case, the CAM) on a different host, you must add the CAM host URL to the API server and the Kubernetes API server CRs.
```
Done. I simplified it and changed title to "4.1 source cluster"

apinnick commented on modules/migration-adding-replication-repository-to-cam.adoc:
```diff
@@ -0,0 +1,28 @@
+
+. Click *Add replication repository* and wait for connection validation.
+
+. Click *Close*. The repository appears in the *Replication repositories* section.
```
Done in other modules.

apinnick commented on modules/migration-adding-replication-repository-to-cam.adoc:
```diff
@@ -0,0 +1,28 @@
+// migration/migrating_openshift_3_to_4/migrating-openshift-3-to-4.adoc
+[id='migration-adding-replication-repository-to-cam_{context}']
+= Adding a replication repository to the CAM web console
+
```
Done. Also for other modules

apinnick commented:
Yes, just found out that link was old. I've updated the file names

ahardin-rh commented on modules/migration-known-issues.adoc:
```diff
@@ -0,0 +1,23 @@
+* Migration of a namespace fails when migrating more than 500 Pods.
+* Migration of multiple namespaces in a single migration plan fails when migrating more than 25 namespaces.
+* When adding an S3 endpoint to the CAM web console, `https://` is supported only for AWS. For other S3 providers, use `http://`.
+* If an AWS bucket is added to the CAM web console and then deleted, its status remains `True` because the MigStorage CR is not updated. (link:https://bugzilla.redhat.com/show_bug.cgi?id=1738564[BZ#1738564])
```
*BZ#1738564*

ahardin-rh commented on modules/migration-understanding-cpma.adoc:
```diff
@@ -0,0 +1,41 @@
++
+[NOTE]
+====
+If you are using an unsigned certificate, you must link:../authentication/certificates/api-server.html#add-named-api-server_api-server-certificates[add the API server certificate manually] to the target cluster.
```
xref:../authentication/certificates/api-server.adoc#add-named-api-server_api-server-certificates[add the API server certificate manually]

ahardin-rh commented on modules/migration-using-cpma.adoc:
```diff
@@ -0,0 +1,99 @@
+* The {product-title} 3 cluster must be updated to the latest synchronous release.
+* The environment health check must be run on the {product-title} 3 cluster to confirm that there are no diagnostic errors or warnings.
+* The {product-title} version must be 3.7, 3.9, 3.10, or 3.11.
+* You must have link:../cli_reference/openshift_cli/administrator-cli-commands.html#policy[`cluster-admin` privileges].
```
xref:../cli_reference/openshift_cli/administrator-cli-commands.adoc#policy[`cluster-admin` privileges].

apinnick commented on migration/migrating-openshift-3-to-4.adoc:
```diff
@@ -0,0 +1,94 @@
+* The source cluster(s) must be {product-title} 3.7 or later.
+* The replication repository object storage must support the S3 API. You can deploy a local S3 provider using the link:https://github.com/fusor/mig-operator/blob/release-1.0/docs/usage/ObjectStorage.md[upstream NooBaa project or AWS S3].
+
+* If your applications use images from the `openshift` namespace, the required versions of the images must be present on the target cluster. If not, you must link:../openshift_images/image-streams-manage.html#images-imagestreams-update-tag_image-streams-managing[update the `imagestreamtags` references] to use an available version that is compatible with your application.
```
Whoops -- thanks. Meant to fix that.

apinnick commented:
@bergerhoffer If I can't use xrefs in modules, can I use footnotes? 

One of the devs wants me to add a line (basically a short procedure) for granting cluster-admin privileges to a user. I told him that a procedure, even a single line, would look weird in a list of prerequisites and he asked whether a link could be provided instead. Since we can't have xrefs in modules, should I put the line in a footnote?

apinnick commented:
@bergerhoffer I can't seem to comment on your comment on `./cpma`, so here's the comment:

> Is "cpma" a directory or a command? The way it's used above, it's an executable file.
> And the "." in "./cpma" means "in the current directory", so it's a relative path.

I did some digging and got more info on exactly what the user is downloading. I also updated the installation procedure because it was missing some details. It's different from our normal installations (like yum or rpm).

`cpma` is the command and is the name of a binary file that the user downloads and saves in a shared binary directory ($PATH for Linux/MacOSX, %PATH% for Windows). That means it can be run from any directory.

`./` in this context is not a path. It indicates that the binary is supposed to run in the current directory, which is *not* defined as $PATH or %PATH%. See https://unix.stackexchange.com/questions/4430/why-do-we-use-to-execute-a-file for more info.

apinnick commented:
@bergerhoffer @ahardin-rh 

I just heard from Marco, the PM, that OCP migration GA has been postponed until middle of next week.

apinnick commented:
@bergerhoffer @ahardin-rh 
Are CRs capitalized or not? I couldn't find them mentioned in any of the guidelines and the docs have both. I'm wondering whether to capitalize 'migration controller'.

bergerhoffer commented:
@apinnick I would check with your devs on the `./` question. `./` denotes that the file you are running is in the current directory.

For example, I have `oc` installed and it is on my path. So I can run `oc --help` from any directory and it will recognize it. 

```
[ahoffer@ahoffer ~] $ oc --help
OpenShift Client

This client helps you develop, build, deploy, and run your applications on any
OpenShift or Kubernetes cluster. It also includes the administrative
commands for managing a cluster under the 'adm' subcommand.
...
```

However, if run `./oc --help`, it fails:

```
[ahoffer@ahoffer ~] $ ./oc --help
bash: ./oc: No such file or directory
```

This is because there is not an `oc` file in that directory. The actual `oc` file is located elsewhere.

So if `cpma` is on the user's path then they shouldn't add the `./` in front. If it's not on the user's path, then they should navigate to the location they downloaded it to and then run `./cpma`.

apinnick commented:
I will check the ./cpma path question

apinnick commented:
@gildub Can you please respond to @bergerhoffer 's [comment](https://github.com/openshift/openshift-docs/pull/17516#issuecomment-545957881)? 

Does the user need to run the `./cpma` command in the same directory as the binary?

bergerhoffer commented:
Sorry @apinnick just saw your question about footnotes. We don't have anything against footnotes in our guidelines. The only thing w/ footnotes is that they go at the very bottom of the entire assembly, not just the module, in case that matters to you.

For example, this module has a table w/ footnotes [1], but it's at the very bottom of the assembly. So it's up to you if that's how you want it rendered.

I would see if it's really necessary to add that procedure. It looks like your assembly already does have a link out on it. I would think that should be enough, and then your modules can just say that you need cluster-admin permissions w/o a link or procedure at all.

[1] http://file.tlv.redhat.com/~apinnick/ocp-migration-new/scalability_and_performance/planning-your-environment-according-to-object-limits.html#cluster-limits_object-limits

apinnick commented:
@bergerhoffer 

Re: link for granting cluster-admin privileges. I'd forgotten that it's also in the assembly. OK, I'll leave it there and remove the links in the modules.

bergerhoffer commented:
@gildub Right, thanks for the details. But really my question is what is right for this specific usage:  (step 2):

http://file.tlv.redhat.com/~apinnick/ocp-migration-new/migration/migrating-openshift-3-to-4.html#migration-using-cpma_migrating-openshift-3-to-4

Because seems like it's either missing a step to say to navigate to the directory where cpma is located, or if it's already on the user's path, then there shouldn't be a `./` in front.

gildub commented:
@bergerhoffer, 

Sorry I missed your point.

Step 2 should be using 
```
$ cpma
```
The `./cpma` is obsolete, likely from an older GitHub version or fusor/cpma when tool was compiled locally.

gildub commented:
@bergerhoffer, 
We're assuming the procedure to install cpma indicates the downloaded program to be stored in common system directory normally used in PATH.

apinnick commented:
$ ./cpma > $ cpma

jwmatthews commented on modules/migration-configuring-cors-3.adoc:
```diff
@@ -0,0 +1,78 @@
++
+----
+corsAllowedOrigins:
+- (?i)//migration-openshift-migration\.apps\.cluster\.com(:|\z) <1>
```
@eriknelson please try these instructions and verify CORS is working.

Marco and I hit issues following these steps when pasting the route as shown above.
When we removed the (?i) and (:|\z) we had success

apinnick replied:
That's strange. I confirmed the instructions for [3.11](https://docs.openshift.com/container-platform/3.11/architecture/infrastructure_components/web_console.html#corsAllowedOrigins).

jwmatthews replied:
@apinnick I went back and confirmed, the instructions are correct for how we specify to configure the CORS.

The issue Marco hit earlier today was that when we ran curl to verify we set the Origin to 
  migration-openshift-migration.apps.cluster.com
When it really should be Origin: https://migration-openshift-migration.apps.cluster.com

I'll add a comment lower in doc to get a little more explicit with what to set for the curl verification

mberube99 commented on modules/migration-configuring-cors-4.adoc:
```diff
@@ -0,0 +1,90 @@
+.Procedure
+
+. Log in to the CAM tool cluster.
+. Obtain the CAM tool URL:
```
I would change:
"Obtain the CAM tool URL:" for "Obtain the CAM tool CORS configuration line:"

I actually thought I was getting a broken REGEX line before realizing it wasn't suppose to return the route, but the exact line config you need with the special characters.

mberube99 commented on modules/migration-configuring-cors-4.adoc:
```diff
@@ -0,0 +1,90 @@
+. Obtain the CAM tool URL:
++
+----
+$ oc get -n openshift-migration route/migration -o go-template='(?i}//{{ .spec.host }}(:|\z){{ println }}' | sed 's,\.,\\.,g'
```
There is a typo with this '(?i}', it should be '(?i)' as per below in the doc.

apinnick replied:
Will fix. This file is not being used in the current release because we are assuming CAM is installed on target cluster.

mberube99 commented on modules/migration-known-issues.adoc:
```diff
@@ -0,0 +1,22 @@
++
+These annotations preserve the UID range, ensuring that the containers retain their file system permissions on the target cluster. There is a risk that the migrated UIDs could duplicate UIDs within an existing or future namespace on the target cluster. (link:https://bugzilla.redhat.com/show_bug.cgi?id=1748440[*BZ#1748440*])
+
+* Restic times out when migrating more than 50 projects. (link:https://bugzilla.redhat.com/show_bug.cgi?id=1749831[*BZ#1749831*])
```
I would remove the 50 projects and 500 pods from the "known issues" section.   
Instead, we need a new section describing the default limit in the tool "10 namespaces" and "100 pods".   And explain that those limits can be change using a the following procedure : @jwmatthews to provide.

jwmatthews commented on modules/migration-configuring-cors-3.adoc:
```diff
@@ -0,0 +1,78 @@
+. Obtain your CAM tool URL:
++
+----
+$ oc get -n openshift-migration route/migration -o go-template='(?i}//{{ .spec.host }}(:|\z){{ println }}' | sed 's,\.,\\.,g'
```
There is a subtle typo here of '}' when we wanted ')', the typo was found in the upstream mig-operator docs and recently updated.

Please change to:
$ oc get -n openshift-migration route/migration -o go-template='(?i)//{{ .spec.host }}(:|\z){{ println }}' | sed 's,\.,\\.,g'


Note, in the original:  go-template='(?i}//
We want:                   go-template='(?i)//

jwmatthews commented on modules/migration-configuring-cors-4.adoc:
```diff
@@ -0,0 +1,90 @@
+"<cluster_url>/apis/migration.openshift.io/v1alpha1/namespaces/openshift-migration/migclusters" \ <1>
+-H "Access-Control-Request-Method: GET" \
+-H "Access-Control-Request-Headers: authorization" \
+-H "Origin: <CAM_tool_url>" <2>
```
-H "Origin: https://<CAM_tool_url>"

Marco and myself lost time today trying to figure out why verification with not working.
Turns out we forgot to include 'http://' or 'https://' in the Origin when we attempted to verify.

The scheme is required based on how we defined the rule in master-config.yml.

jwmatthews commented on modules/migration-configuring-cors-3.adoc:
```diff
@@ -0,0 +1,78 @@
+"<cluster_url>/apis/migration.openshift.io/v1alpha1/namespaces/openshift-migration/migclusters" \ <1>
+-H "Access-Control-Request-Method: GET" \
+-H "Access-Control-Request-Headers: authorization" \
+-H "Origin: <CAM_tool_url>" <2>
```
-H "Origin: https://<CAM_tool_url>"

Marco and myself lost time today trying to figure out why verification with not working.
Turns out we forgot to include 'http://' or 'https://' in the Origin when we attempted to verify.

The scheme is required based on how we defined the rule in master-config.yml.

jwmatthews commented on modules/migration-configuring-cors-3.adoc:
```diff
@@ -0,0 +1,78 @@
+.Procedure
+
+. Log in to the {product-title} 4 cluster.
+. Obtain your CAM tool URL:
```
Comments from Marco (he left them originally on the CORS 4 configuration page)

I would change:
"Obtain the CAM tool URL:" for "Obtain the CAM tool CORS configuration line:"

I actually thought I was getting a broken REGEX line before realizing it wasn't suppose to return the route, but the exact line config you need with the special characters.

apinnick replied:
Yes, that does make a difference. I think that it would be a good idea for me to remove the note explaining the syntax because the user just has to copy the CORS config line.

jwmatthews commented on modules/migration-configuring-cors-4.adoc:
```diff
@@ -0,0 +1,90 @@
+[source,yaml]
+----
+spec:
+  unsupportedConfigOverrides:
```
Note this block of config is for the 4.1 side, not needed on 4.2.

Refer to upstream docs here on how to configure OCP 4.2 or OCP 4.1
https://github.com/fusor/mig-operator/blob/master/README.md#openshift-4

apinnick commented on modules/migration-creating-migration-plan-cam.adoc:
```diff
@@ -0,0 +1,42 @@
+. Select a *Source cluster*.
+. Select a *Target cluster*.
+. Select a *Replication repository*.
+. Select the projects to be migrated and click *Next*.
```
@jwmatthews 

Why does the UI say "Select projects to be migrated" when we call them "namespaces" elsewhere in the documentation?

jwmatthews replied:
OpenShift term is 'projects'
Kubernetes calls them 'namespaces'

Terms are used interchangeably for most part, some times there are subtle reasons why it must be called a project, but for most part we use then interchangeably.

Note upstream tools like Velero will always refer to namespaces.

apinnick replied:
OK, I will leave them as they are.

apinnick commented on modules/migration-changing-migration-plan-limits.adoc:
```diff
@@ -0,0 +1,26 @@
+
+You can change the default limits by updating the following values in the Migration controller CR:
+
+[source,yaml]
```
@jwmatthews 

Is this sufficient or do we need to include a procedure for editing the CR yaml? I don't have time to set up a 4.2 cluster today.

jwmatthews replied:
I think below is sufficient just showing the values to change.

If you wanted to show how to change this, an example would be:

$ oc get migrationcontroller -n openshift-migration
NAME                   AGE
migration-controller   5d19h

$ oc edit migrationcontroller -n openshift-migration

apinnick commented on modules/migration-adding-replication-repository-to-cam.adoc:
```diff
@@ -0,0 +1,44 @@
++
+[NOTE]
+====
+You can deploy a local S3 object storage with the link:https://github.com/fusor/mig-operator/blob/release-1.0/docs/usage/ObjectStorage.md[upstream NooBaa project or AWS S3].
```
@jwmatthews 

Was the final decision to keep this link to github and the upstream NooBaa project?

jwmatthews replied:
Yes, final decision is to continue with this github link.

This is a temporary path, for next ~month or so we will continue to point to this upstream documentation.
Once OCS is officially released we plan to switch to using downstream NooBaa

apinnick commented on modules/migration-changing-migration-plan-limits.adoc:
```diff
@@ -0,0 +1,26 @@
+# cluster where `migration_controller: true`
+mig_pv_limit: 100
+mig_pod_limit: 100
+mig_namespace_limit: 10
```
@jwmatthews 

Can we provide any guidelines for max PVs and Pods? 

Also, if exceeding the namespace limit produces an error message and the user can't create migration plan, why are we allowing the user to change the limit here?

jwmatthews replied:
Little scale testing has been done up to now.  We have seen success of 1000 namespace migrations.

The reasoning for these limits is to reduce chances of customers hitting issues at scale unexpectedly.  The limits are kind of a communication aide to be cautious.

I don't think we have much more guidance on the numbers to share.

We chose a limit of 100 PVs and 100 Pods as an arbitrary limit we felt would be successful.  Scale testing is focused on confirming these, but we don't expect to have that data for another ~2 weeks.

I would assume as we learn more and enhance scale issues these numbers will be increased in future.

mberube99 commented on modules/migration-changing-migration-plan-limits.adoc:
```diff
@@ -0,0 +1,87 @@
+[id='migration-changing-migration-plan-limits_{context}']
+= Changing migration plan limits for large migrations
+
+You can change the migration plan limits for large migrations.
```
Can we have something around those lines:

"This tool has been pre-configured with some migration limits.   Those limits are there as safety guards to reduce risk during your migration process.  We recommend changing those limits only after they have been tested in your environment."

apinnick replied:
I've added an "IMPORTANT" note about first testing changes in their environment, following up on our gchat.

XinRedhat commented:
@apinnick 
Would you please clarify that the update happens while creating Migration Controller? 
>Update the following values in the Migration controller CR:

For the error/critical message of migration plan limits, they will appear in the page. This feature already is implemented in PR https://github.com/fusor/mig-ui/pull/604

apinnick commented:
> @apinnick
> Would you please clarify that the update happens while creating Migration Controller?
> 
> > Update the following values in the Migration controller CR:

Why do they have to change the parameters while creating the Migration controller? Why can't they edit the CR?

> For the error/critical message of migration plan limits, they will appear in the page. This feature already is implemented in PR [fusor/mig-ui#604](https://github.com/fusor/mig-ui/pull/604)

Yes, I just received the screenshots. I'm going to use the UI messages instead of the CR content because we can assume that the user is more likely to see the warnings in the UI than in the CR.

bergerhoffer commented:
@apinnick Is Thursday still the GA date for this? And are you finished with additional updates? I can try to run through it again now, but we need to leave time to get a final peer review, and address feedback from that, after any other changes are made.

In the future, it would be good to submit the initial content, and get it peer reviewed/merged, and then follow up in separate PRs for additional updates. Having one large PR makes it more difficult to peer review and keep up with the changes. And runs into the risk of it not being ready to go out the day it needs to, since we need to do a final peer review before merging.

What other teams have been doing when adding content that isn't ready to be published is adding the content, but commenting out the new files in the _topic_map.yml file. So I'd recommend that for next time.

bergerhoffer commented:
As of the changes right now, this looks good to me.

eriknelson commented on modules/migration-installing-migration-operator-manually.adoc:
```diff
@@ -0,0 +1,70 @@
+// * {product-title} 3:
++
+----
+$ podman cp $(podman create registry-redhat.io/rhcam/openshift-migration-rhel7-operator:v1.0 ):/controller-3.yml ./
```
`registry-redhat.io` should be `registry.redhat.io`

eriknelson replied:
rhcam here should be rhcam-1-0, full command:

`podman cp $(podman create registry.redhat.io/rhcam-1-0/openshift-migration-rhel7-operator:v1.0 ):/controller-3.yml ./`

eriknelson commented on modules/migration-installing-migration-operator-manually.adoc:
```diff
@@ -0,0 +1,70 @@
+. Download the `operator.yml` file:
++
+----
+$ podman cp $(podman create registry.redhat.io/rhcam/openshift-migration-rhel7-operator:v1.0 ):/operator.yml ./
```
`rhcam` here should be `rhcam-1-0`, full command:

`podman cp $(podman create registry.redhat.io/rhcam-1-0/openshift-migration-rhel7-operator:v1.0 ):/operator.yml ./`

apinnick commented:
@bergerhoffer This PR is ready for final review and merge. Thanks.

bergerhoffer commented:
LGTM, merging!

bergerhoffer commented:
/cherrypick enterprise-4.2

bergerhoffer commented:
/cherrypick enterprise-4.3
